src/beast/beast/asio/buffers_readstream.h

```diff
@@ -0,0 +1,147 @@
+    consuming_buffers<ConstBufferSequence> cb_;
+
+public:
+    /// The type of the next layer.
+    using next_layer_type =
+        std::remove_reference_t<Stream>;
+
+    /// The type of the lowest layer.
+    using lowest_layer_type =
+        typename next_layer_type::lowest_layer_type;
+
+    template<class... Args>
+    explicit
+    buffers_readstream(
+        ConstBufferSequence const& buffers, Args&&... args);
+
+    /// Get a reference to the next layer.
+    next_layer_type&
+    next_layer()
+    {
+        return next_layer_;
+    }
+
+    /// Get a reference to the lowest layer.
+    lowest_layer_type&
+    lowest_layer()
+    {
+        return next_layer_.lowest_layer();
+    }
+
+    /// Get a const reference to the lowest layer.
+    lowest_layer_type const&
+    lowest_layer() const
+    {
+        return next_layer_.lowest_layer();
+    }
+
+    /// Get the io_service associated with the object.
+    boost::asio::io_service&
+    get_io_service()
+    {
+        return next_layer_.get_io_service();
+    }
+
+    /** Returns the unused portion of the input sequence. */
+    auto const&
+    data() const
+    {
+        return cb_;
+    }
+
+    /// Write the given data to the stream. Returns the number of bytes written.
+    /// Throws an exception on failure.
+    template<class OtherConstBufferSequence>
+    std::size_t
+    write_some(OtherConstBufferSequence const& buffers)
+    {
+        return next_layer_.write_some(buffers);
+    }
+
+    /// Write the given data to the stream. Returns the number of bytes written,
+    /// or 0 if an error occurred.
+    template <class OtherConstBufferSequence>
+    std::size_t
+    write_some(OtherConstBufferSequence const& buffers,
+        error_code& ec)
+    {
+        return next_layer_.write_some(buffers, ec);
+    }
+
+    /// Start an asynchronous write. The data being written must be valid for the
+    /// lifetime of the asynchronous operation.
+    template<class OtherConstBufferSequence, class WriteHandler>
+    auto
+    async_write_some(OtherConstBufferSequence const& buffers,
+        WriteHandler&& handler);
+
+    /// Read some data from the stream. Returns the number of bytes read.
+    /// Throws an exception on failure.
+    template<class MutableBufferSequence>
+    std::size_t
+    read_some(MutableBufferSequence const& buffers);
+
+    /// Read some data from the stream. Returns the number of bytes read
+    /// or 0 if an error occurred.
+    template<class MutableBufferSequence>
+    std::size_t
+    read_some(MutableBufferSequence const& buffers,
+        error_code& ec);
+
+    /// Start an asynchronous read. The buffer into which the data will be read
+    /// must be valid for the lifetime of the asynchronous operation.
+    template<class MutableBufferSequence, class ReadHandler>
+    auto
+    async_read_some(MutableBufferSequence const& buffers,
+        ReadHandler&& handler);
+};
+
+} // beast
+
+#include <beast/asio/impl/buffers_readstream.ipp>
```

Comment:
Style nit:  I'm not a big fan of this use of "implementation headers".  It hurts compilation time because the compiler has to open another file.  And it hurts readability because the reader has to open another file to get the rest of the template definition.  I can't think of anything it improves.  That being said, a lot of people sure seem to like this idiom.